tensorflow_serving/sources/storage_path/file_system_storage_path_source.cc

```diff
@@ -76,34 +77,50 @@ std::set<string> GetDeletedServables(
   return deleted_servables;
 }
 
-// Like PollFileSystemForConfig(), but for a single servable.
-Status PollFileSystemForServable(
+void EmitAllVersions(
     const FileSystemStoragePathSourceConfig::ServableToMonitor& servable,
-    std::vector<ServableData<StoragePath>>* versions) {
-  // First, determine whether the base path exists. This check guarantees that
-  // we don't emit an empty aspired-versions list for a non-existent (or
-  // transiently unavailable) base-path. (On some platforms, GetChildren()
-  // returns an empty list instead of erring if the base path isn't found.)
-  if (!Env::Default()->FileExists(servable.base_path())) {
-    return errors::InvalidArgument("Could not find base path ",
-                                   servable.base_path(), " for servable ",
-                                   servable.servable_name());
-  }
+    std::vector<ServableData<StoragePath>>* versions,
+    const std::vector<string>& children) {
 
-  // Retrieve a list of base-path children from the file system.
-  std::vector<string> children;
-  TF_RETURN_IF_ERROR(
-      Env::Default()->GetChildren(servable.base_path(), &children));
+  LOG(INFO) << "Emit all servable versions";
+
+  // Identify all the versions, among children that can be interpreted as
+  // version numbers.
+  int version_child = -1;
+  int64 found_version;
+  bool model_aspired = false;
 
-  // GetChildren() returns all descendants instead for cloud storage like GCS.
-  // In such case we should filter out all non-direct descendants.
-  std::set<string> real_children;
   for (int i = 0; i < children.size(); ++i) {
     const string& child = children[i];
-    real_children.insert(child.substr(0, child.find_first_of('/')));
+    int64 child_version_num;
+    if (!strings::safe_strto64(child.c_str(), &child_version_num)) {
+      continue;
+    }
+
+    version_child = i;
+    found_version = child_version_num;
+    // Emit all the aspired-versions data.
+    if (version_child >= 0) {
+      const ServableId servable_id = {servable.servable_name(), found_version};
+      const string full_path =
```

Comment:
Can we avoid duplicating code between EmitAllVersions() and EmitLatestVersion()?

Maybe have one function that decides which versions to emit (just which 'children' entries), and then a separate one that converts them into ServableData items.